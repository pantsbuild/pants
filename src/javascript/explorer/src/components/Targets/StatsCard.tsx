import { useTheme } from '@mui/material/styles';
import { darken, alpha, lighten } from '@mui/system';
import useMediaQuery from '@mui/material/useMediaQuery';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
//import Divider from '@mui/material/Divider';
import Stack from '@mui/material/Stack';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Typography from '@mui/material/Typography';
import Masonry from '@mui/lab/Masonry';

import { useTargets } from "lib/target-data";
import Loading from "components/Loading";
import Error from "components/Error";
import RefetchButton from "components/RefetchButton";
import { StatsCardProps } from './types';


type _Counts = {
  [key: string]: number
};


export const StatsCard = ({ targets }: StatsCardProps) => {
  const theme = useTheme();
  const target_types_counts: _Counts = targets.reduce(
    (counts, target) => {
      if (target !== undefined) {
        counts[target.targetType] = (counts[target.targetType] || 0) + 1;
      }
      return counts;
    },
    {} as _Counts
  );

  const target_types_content = Object.entries(target_types_counts).sort().map(
    ([target_type, count], index) => (
      <TableRow key={index}>
        <TableCell>{target_type}</TableCell>
        <TableCell align="right">{count}</TableCell>
      </TableRow>
    )
  );

  const border_color = theme.palette.mode === 'light'
                     ? lighten(alpha(theme.palette.divider, 1), 0.88)
                     : darken(alpha(theme.palette.divider, 1), 0.68);
  const sum_row_border = `4px double ${border_color}`;

  return (
    <Card>
      <CardContent>
        <Typography variant="h6">
          Target type distribution across your repository.
        </Typography>
        <TableContainer>
          <Table size="small">
            <TableHead>
              <TableRow>
                <TableCell>Target</TableCell>
                <TableCell align="right">Count</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {target_types_content}
              <TableRow>
                <TableCell sx={{ borderTop: sum_row_border }}>Total</TableCell>
                <TableCell sx={{ borderTop: sum_row_border }} align="right">{targets.length}</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        </TableContainer>
      </CardContent>
    </Card>
  );
};


export default () => {
  const [targets, loading, error] = useTargets(["address", "targetType"]);

  if (loading) {
    return <Loading title="Loading targets data..." />;
  }

  if (error) {
    return <Error title="Failed to load target data" error={error.toString()} />;
  }

  return <StatsCard targets={targets} />
}
