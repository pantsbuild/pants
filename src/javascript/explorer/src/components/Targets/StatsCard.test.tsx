import { act, default as renderer, ReactTestRenderer } from 'react-test-renderer';
import { StatsCard } from './StatsCard';


test('StatsCard renders table.', async () => {
  const targets = [
    {address: "//.gitignore:files", targetType: "file", fields: {}},
    {address: "//BUILD_ROOT:files", targetType: "file", fields: {}},
    {address: "//cargo:scripts", targetType: "shell_source", fields: {}},
    {address: "//pants:scripts", targetType: "shell_source", fields: {}},
    {address: "//pants.toml:files", targetType: "file", fields: {}},
    {address: "3rdparty/jvm/com/fasterxml/jackson/core:jackson-databind", targetType: "jvm_artifact", fields: {}},
    {address: "3rdparty/jvm/io/circe:circe-generic", targetType: "jvm_artifact", fields: {}},
    {address: "3rdparty/jvm/org/scalameta:scalameta", targetType: "jvm_artifact", fields: {}},
    {address: "3rdparty/python#PyYAML", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#ansicolors", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#chevron", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#fastapi", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#fasteners", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#freezegun", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#humbug", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#ijson", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#mypy-typing-asserts", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#packaging", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#pex", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#psutil", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#pytest", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#python-lsp-jsonrpc", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#requests", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#setproctitle", targetType: "python_requirement", fields: {}},
    {address: "3rdparty/python#setuptools", targetType: "python_requirement", fields: {}},
  ];

  let component;
  await act(async () => {
    component = renderer.create(
      <StatsCard targets={targets} />
    );
  });

  let tree = (component as unknown as ReactTestRenderer).toJSON();
  expect(tree).toMatchSnapshot();
});
