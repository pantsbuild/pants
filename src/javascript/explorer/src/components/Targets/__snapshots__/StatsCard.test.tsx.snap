// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsCard renders table. 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <h6
      className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
    >
      Target type distribution across your repository.
    </h6>
    <div
      className="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
    >
      <table
        className="MuiTable-root css-rqglhn-MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
              scope="col"
            >
              Target
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall css-6qfsqn-MuiTableCell-root"
              scope="col"
            >
              Count
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              file
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
            >
              3
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              jvm_artifact
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
            >
              3
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              python_requirement
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
            >
              17
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
            >
              shell_source
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
            >
              2
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1pigulk-MuiTableCell-root"
            >
              Total
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-1m2cc54-MuiTableCell-root"
            >
              25
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
